content/backend/typescript-apollo/6-authentication.md

```diff
@@ -0,0 +1,776 @@
+    userId: number;
+}
+
+export function decodeAuthHeader(authHeader: String): AuthTokenPayload { // 2
+    const token = authHeader.replace("Bearer ", "");  // 3
+
+    if (!token) {
+        throw new Error("No token found");
+    }
+    return jwt.verify(token, APP_SECRET) as AuthTokenPayload;  // 4
+}
+```
+
+</Instruction>
+
+Let's understand what is going on here: 
+
+1. `AuthTokenPayload` interface is based on the shape of the JWT token that we issued during `signup` and `login`. When the server decodes an issued token, it should expect a response in this format. 
+
+2. The `decodeAuthHeader` function takes the `Authorization` header and parses it to return the payload of the JWT. 
+
+3. The `Authorization` header, contains the `type` or scheme of authorization followed by the token. In our case, `"Bearer"` represents the authorization scheme. Since the server is only interested in the JWT token itself, you can get rid of the `"Bearer"` and keep only the token. _If you're interested to learn more, here is a nice [stackoverflow question](https://security.stackexchange.com/q/108662) explaining this convention_. 
+
+4. The `jwt.verify()` functions decodes the token. It also needs access to the secret (or a public key) which was used to sign the token. 
+
+
+
+There's one more issue to solve. The GraphQL resolvers don't have access to the `request` headers and have no way to get the token from the `Authorization` header. To solve this, you are going to add the `userId` received from the decoded JWT to the  GraphQL `context`, so that every resolver has access to the `userId`. 
+
+<Instruction>
+
+Update `src/context.ts` to add the decoded JWT from the request header:
+
+
+```typescript{2-3,9,12-22}(path=".../hackernews-typescript/src/context.ts")
+import { PrismaClient } from "@prisma/client";
+import { decodeAuthHeader, AuthTokenPayload } from "./utils/auth";   
+import { Request } from "express";  
+
+export const prisma = new PrismaClient();
+
+export interface Context {
+    prisma: PrismaClient;
+    userId?: number;  // 1
+}
+
+export const context = ({ req }: { req: Request }): Context => {   // 2
+    const token =
+        req && req.headers.authorization
+            ? decodeAuthHeader(req.headers.authorization)
+            : null;
+
+    return {  
+        prisma,
+        userId: token?.userId, 
+    };
+};
+```
+
+</Instruction>
+
+
+Let's understand the changes made to `context.ts`: 
+
+
+1. The context `interface` is updated to have a `userId` type. This is optional because no `userId` will be attached to the `context` when requests are sent without the `Authorization` header. 
+
+2. Instead of being an object, `context` is now a _function_ which needs to be executed to return the actual object of type `Context`. Apollo-server is smart enough to recognize this change from object to function and will execute the function (with some arguments, like the HTTP request) to resolve the final context object. 
```

Comment:
```suggestion
2. Instead of being an object, `context` is now a _function_ which needs to be executed to return the actual object of type `Context`. Apollo Server is smart enough to recognize this change from object to function and will execute the function (with some arguments, like the HTTP request) to resolve the final context object. 
```